pandas/tests/frame/methods/test_diff.py

```diff
@@ -302,3 +302,10 @@ def test_diff_all_int_dtype(self, any_int_numpy_dtype):
         )
         expected = DataFrame([np.nan, 1.0, 1.0, 1.0, 1.0], dtype=expected_dtype)
         tm.assert_frame_equal(result, expected)
+
+    def test_diff_axis1_all_bool_dtype(self):
+        # https://github.com/pandas-dev/pandas/issues/52579
+        df = DataFrame({"col": [True, False]})
+        result = df.T.diff(axis=1)
+        expected = DataFrame({0: [np.nan], 1: [True]}, index=["col"])
```

Comment:
this has float dtype for expected[0].  is that what we want?  id have expected object

Reply:
For this test, the expected value for `df.T.diff(axis=1)` is

```
	0	1
col	NaN	True
```.

This is because I think it should be consistent with diff() results in other cases as shown below.

code:
```
df = pd.DataFrame({'a': [1, 2, 3, 4, 5, 6],
                                'b': [1, 1, 2, 3, 5, 8],
                                'c': [1, 4, 9, 16, 25, 36]})
result = df4.T.diff(axis=1)
print(result)
```

result:
```
            a  b   c
        0 NaN  0   0
        1 NaN -1   3
        2 NaN -1   7
        3 NaN -1  13
        4 NaN  0  20
        5 NaN  2  28
```

Reply:
im not asking about the _values_, but the dtypes

Reply:
First of all, I needed to correct the result of the above code as below. This could have caused confusion.

```
     a    b     c
0  NaN  NaN   NaN
1  1.0  0.0   3.0
2  1.0  1.0   5.0
3  1.0  1.0   7.0
4  1.0  2.0   9.0
5  1.0  3.0  11.0
```

And if you check the dtype of result[0] in the above code, you see float64 instead of the object as follows.

code:
```
print(result.dtypes)
```

result:
```
0    float64
1      int64
2      int64
3      int64
4      int64
5      int64
dtype: object
```

So I think the dtype of 'expected[0]' should be float64 to be consistent with the above results.

Reply:
@mroeschke am i wrong to expect object dtype here?

Reply:
Do you expect object because even though for axis=1 a value might not have a corresponding diff, the result dtype should still respect the original `bool` type with `float` nan?

Reply:
bc if you do the diff on axis=0 you get object

Reply:
Well for axis=1 the nans (value without a corresponding diff) are in their own column so `float` could be a more useful data type here. Plus `df.T.diff(axis=1).T == df.diff(axis=0)` should still hold